nodejs-server/src/echoServer.ts

```diff
@@ -0,0 +1,316 @@
+import * as grpc from 'grpc';
+import * as Long from 'long';
+
+import * as echoTypes from '../pbjs-genfiles/echo.js';
+
+import showcaseV1Beta1 = echoTypes.google.showcase.v1beta1;
+import google = echoTypes.google;
+import longrunning = google.longrunning;
+import {OperationsServer} from './operationsServer.js';
+
+/**
+ * Implements Echo server based on echo.proto.
+ */
+export class EchoServer {
+  private requestCount: number;
+  private paginationOutput: {[index: string]: string[]};
+  private paginationRequest: {[index: string]: string|null|undefined};
+  private operationsServer: OperationsServer;
+  private verbose: boolean;
+
+  constructor(verbose: boolean, operationsServer: OperationsServer) {
+    this.requestCount = 0;
+    this.paginationOutput = {};
+    this.paginationRequest = {};
+    this.operationsServer = operationsServer;
+    this.verbose = verbose;
+  }
+
+  private log(request: number, ...args: Array<string|{}>) {
+    if (this.verbose) {
+      console.log(`[EchoServer request #${request}]`, ...args);
+    }
+  }
+
+  /**
+   * Helper function to extract `google.rpc.Status` from some requests that
+   * contain this field.
+   */
+  private static requestToStatus(request:
+                                     showcaseV1Beta1.EchoRequest|
+                                 showcaseV1Beta1.ExpandRequest|
+                                 showcaseV1Beta1.WaitRequest):
+      google.rpc.Status {
+    const message = request.error && request.error.message ?
+        request.error.message :
+        'Error';
+    const code = request.error && request.error.code ?
+        request.error.code :
+        grpc.status.INVALID_ARGUMENT;
+    const details =
+        request.error && request.error.details ? request.error.details : [];
+    const error = new google.rpc.Status();
+    error.code = code;
+    error.message = message;
+    error.details = details;
+    return error;
+  }
+
+  /**
+   * Helper function to build `grpc.ServiceError` out of `google.rpc.Status`.
+   */
+  private static requestToError(request:
+                                    showcaseV1Beta1.EchoRequest|
+                                showcaseV1Beta1.ExpandRequest|
+                                showcaseV1Beta1.WaitRequest):
+      grpc.ServiceError {
+    const status = EchoServer.requestToStatus(request);
+    const error = new Error(status.message) as grpc.ServiceError;
+    error.code = status.code;
+    return error;
+  }
+
+  /**
+   * Unary call example. Echoes the request.
+   */
+  echo(
```

Comment:
Just style, but use async instead of callback based

Reply:
(copying from another comment thread) I think (call, callback) => void is the interface that gRPC expects. This EchoServer class is being fed to grpc.Server.addService() and they expect certain API.

Reply:
maybe not worth it, but could we wrap the `grpc.Server.addService` method, such that downstream libraries can use `async`, e.g.,

```js
addService(service) {
  grpc.Server.addService(util.callbackify(service))
}
```

maybe not worth it for this PR?